modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C

```diff
@@ -0,0 +1,635 @@
+    };
+    TIMPI::push_parallel_vector_data(_communicator, push_data, action_functor);
+  }
+
+  // Then pull
+  {
+    auto gather_functor = [this](const processor_id_type libmesh_dbg_var(pid),
+                                 const std::vector<dof_id_type> & elem_ids,
+                                 std::vector<VectorValue<ADReal>> & data_to_fill) {
+      mooseAssert(pid != this->processor_id(), "We shouldn't be gathering from ourselves.");
+      data_to_fill.resize(elem_ids.size());
+      for (const auto i : index_range(elem_ids))
+      {
+        const auto id = elem_ids[i];
+        auto it = _a.find(id);
+        mooseAssert(it != _a.end(), "We should hold the value for this locally");
+        data_to_fill[i] = it->second;
+      }
+    };
+
+    auto action_functor = [this](const processor_id_type libmesh_dbg_var(pid),
+                                 const std::vector<dof_id_type> & elem_ids,
+                                 const std::vector<VectorValue<ADReal>> & filled_data) {
+      mooseAssert(pid != this->processor_id(), "The requst filler shouldn't have been ourselves");
+      mooseAssert(elem_ids.size() == filled_data.size(), "I think these should be the same size");
+      for (const auto i : index_range(elem_ids))
+      {
+        const auto id = elem_ids[i];
+        auto it = _a.find(id);
+        mooseAssert(it != _a.end(), "We requested this so we must have it in the map");
+        it->second = filled_data[i];
+      }
+    };
+    TIMPI::pull_parallel_vector_data(
+        _communicator, pull_requests, gather_functor, action_functor, &example);
+  }
+#else
+  mooseError("INSFVRhieChowInterpolator only supported for global AD indexing.");
+#endif
+}
+
+void
+INSFVRhieChowInterpolator::computeFirstAndSecondOverBars()
+{
+  _b2.reserve(_fe_problem.getEvaluableElementRange().size());
+
+  if (_standard_body_forces)
+  {
+    for (const auto & pr : _b)
+      _b2[pr.first] = pr.second;
+
+    return;
+  }
+
+  Moose::FV::interpolateReconstruct(_b2, _b, 1, false, _evaluable_fi, *this);
+}
+
+void
+INSFVRhieChowInterpolator::applyBData()
+{
+  const auto s = _sys.number();
+  for (auto * const elem : *_elem_range)
+  {
+    const auto elem_volume = _assembly.elementVolume(elem);
+    for (const auto i : index_range(_var_numbers))
+    {
+      const auto vn = _var_numbers[i];
+      // negative here because we swapped the sign in addToB and so now we need to swap it back
+      const auto residual = -elem_volume * _b2[elem->id()](i);
+      const auto dof_index = elem->dof_number(s, vn, 0);
+
+      if (_fe_problem.currentlyComputingJacobian())
+        _assembly.processDerivatives(residual, dof_index, _matrix_tags);
+      else
+        _assembly.processResidual(residual.value(), dof_index, _vector_tags);
+    }
+  }
+
+  if (_fe_problem.currentlyComputingJacobian())
+    _assembly.addCachedJacobian();
+  else
+    _assembly.addCachedResiduals();
+}
+
+void
+INSFVRhieChowInterpolator::finalizeBData()
```

Comment:
and this is to know what the _b terms are around the element

Reply:
Yes. We need to be sure that we have two ghost layers of element evaluations of B, or else we will not have accurate computation of the third overbar of B.